Add note about error message style

ddn0 opened the pull request:
Too much consistency may be bad but some consistency in the face of
ambiguity is usually better than nothing.

(For context see https://github.com/KatanaGraph/katana-enterprise/pull/934#discussion_r600783468)

ddn0 commented on libgalois/include/galois/OutIndexView.h:
```diff
@@ -34,7 +34,7 @@ struct GRPrefix {
 
 class OutIndexView {
-  tsuba::FileView file_;
+  tsuba::file::FileView file_;
```
I think the goal of clearly delineating the different levels of the API is a good one, but I don't know if sub namespaces is best way to get at this idea for the reasons raised here: https://abseil.io/tips/130

I also don't know how much we need to discourage people (via coding idioms) from reading and writing to S3 directly. It's pretty painful to write the proper batch loading/async code so most people will want to use the structured interfaces where possible anyway.

tylershunt replied:
My initial plan was just to change these functions to have the suffix `File` is that preferable?

And oh sure! I only meant that it should look different to write files vs writing graphs.

tylershunt replied:
I ended up changing the name of unstructured files to `FileOpen, FileStore...` to match FileView.

ddn0 commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -30,6 +33,40 @@ struct RDG {
+galois::Result<void> Rename(std::shared_ptr<RDGHandle> handle,
+                            const std::string& name, int flags);
+
+// Load the RDG described by the metadata in @handle into memory
```
if you add `///` or similar, this comment should show up in our doxygen docs.

ddn0 commented on libtsuba/src/file.cpp:
```diff
@@ -191,67 +180,85 @@ static int S3OrDoFile(const std::string& uri, const uint8_t* data,
+
+namespace tsuba {
+
+int FileOpen(const std::string& uri) {
```
probably drop `namespace tsuba` and define as `tsuba::FileOpen`? This makes sure that FileOpen the definition matches the declaration. Otherwise, typos create linker errors rather than compiler ones.

https://godbolt.org/z/y-uxqr

tylershunt replied:
Oh nice; I didn't realize that there was a difference.

witchel commented on libsupport/src/FileSystem.cpp:
```diff
@@ -34,7 +34,7 @@ galois::OpenUniqueFile(std::string_view prefix, std::string_view suffix) {
 
   int fd = mkstemps(buf.data(), suffix.length());
   if (fd < 0) {
-    return std::error_code(errno, std::system_category());
```
Oh, `std::system_category() ` is a new one for me.  Yeah, good idea.

tylershunt replied:
👍

witchel commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -30,6 +33,40 @@ struct RDG {
+// acceptable values for Rename's flags
+constexpr int kOverwrite = 1;
+// Rename
+galois::Result<void> Rename(std::shared_ptr<RDGHandle> handle,
```
I saw Rename and thought it had something to do with storage.  But it looks like it just changes the graph name.  Seems fine.  I wonder if we will eventually want a general "ChangeMetadata" type call, but let's let that future happen if it wants to happen.

tylershunt replied:
That's the eventual intent, i'm using it here to also change the names of graphs that are only in memory; That's changing soon but this is just the code that I had before moved over.

witchel commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -30,6 +33,40 @@ struct RDG {
+galois::Result<void>
+AddNodeProperties(RDG* rdg, const std::shared_ptr<arrow::Table>& table);
+galois::Result<void>
+AddEdgeProperties(RDG* rdg, const std::shared_ptr<arrow::Table>& table);
```
So we only see properties, not labels, right?  I know there was debate about that with the graph folks because Neo4j distinguishes.

tylershunt replied:
Yes, afaik labels are just properties with extra metadata to us.

witchel commented on libtsuba/src/file.cpp:
```diff
@@ -155,29 +160,13 @@ class MappingDesc {
-static int S3OrDoFile(const std::string& uri, const uint8_t* data,
-                      uint64_t size) {
-  if (IsUri(uri)) {
+int S3OrDoFile(const std::string& uri, const uint8_t* data, uint64_t size) {
```
Is this function now exported?

tylershunt replied:
I moved the file local symbols into an anonymous namespace `namespace {...}`. It's achieves the same thing as `static` but is less `c` and more `c++` plus other new code is doing things that way so I thought it would be good to be consistent.

witchel replied:
Oh, I think my mother occasionally put my father in an anonymous namespace.  I just didn't know that was what it was called at the time.

ddn0 replied:
BTW, I think something we might want to do soon is to hide symbols by default and explicitly mark symbols we want to export: https://github.com/IntelligentSoftwareSystems/Galois/issues/303

ddn0 committed:
Add note about error message style

Too much consistency may be bad but some consistency in the face of
ambiguity is usually better than nothing.

arthurp commented on CONTRIBUTING.md:
```diff
@@ -500,6 +500,58 @@ it is acceptable to follow the convention in that module; though in general, it
 than you found it."
 
+## Error Codes and Messages
+
```
Since this is being added, how about we also document the general approach to error handling to help people who are not already aware of how `boost::outcome` or sum-type based error handling in general works? 

Here we can have links to `katana::Result` and `KATANA_ERROR`. And in the doc comments of those they can have examples of how they are used and links to `boost::outcome` for more details.

witchel commented on CONTRIBUTING.md:
```diff
@@ -500,6 +500,58 @@ it is acceptable to follow the convention in that module; though in general, it
+    int u = n / number;
+  } else {
+    return KATANA_ERROR(IllegalArgument, "number should be positive");
+  }
```
Argument is an int, so "should be non-negative"?

ddn0 replied:
maybe I was being too obscure, but I was trying to convey non-zero as well (zero being neither negative nor positive).

tylershunt reviewed (approved):
LGTM

arthurp commented on CONTRIBUTING.md:
```diff
@@ -500,20 +500,37 @@ it is acceptable to follow the convention in that module; though in general, it
+
+  // ... we handle it
+  if (r.error() == ErrorCode::Foo) {
+    continue;
```
Don't forget `continue` is a keyword and not valid here. Do you mean `// ... continue`, as in "code that implements handling goes here"?

ddn0 replied:
I updated the PR to something more believable.
